File: 3.x.x/crates/utopia-lock/src/semaphore.rs
use std::time::{Duration, Instant};

use parking_lot::{Condvar, Mutex as ParkingMutex};

use crate::error::Contention;
use crate::lock::Lock;

#[derive(Debug)]
struct Inner {
    held: usize,
}

/// PHP `Utopia\Lock\Semaphore`.
#[derive(Debug)]
pub struct Semaphore {
    permits: usize,
    inner: ParkingMutex<Inner>,
    cond: Condvar,
}

impl Semaphore {
    pub fn new(permits: usize) -> Result<Self, String> {
        if permits < 1 {
            return Err("Permits must be at least 1".into());
        }
        Ok(Self {
            permits,
            inner: ParkingMutex::new(Inner { held: 0 }),
            cond: Condvar::new(),
        })
    }
}

impl Lock for Semaphore {
    fn acquire(&self, timeout: f64) -> bool {
        let mut guard = self.inner.lock();
        if guard.held < self.permits {
            guard.held += 1;
            return true;
        }
        if timeout == 0.0 {
            return false;
        }
        if timeout < 0.0 {
            while guard.held >= self.permits {
                self.cond.wait(&mut guard);
            }
            guard.held += 1;
            return true;
        }
        let deadline = Instant::now() + Duration::from_secs_f64(timeout);
        while guard.held >= self.permits {
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return false;
            }
            if self.cond.wait_for(&mut guard, remaining).timed_out() && guard.held >= self.permits {
                return false;
            }
        }
        guard.held += 1;
        true
    }

    fn try_acquire(&self) -> bool {
        self.acquire(0.0)
    }

    fn release(&self) {
        let mut guard = self.inner.lock();
        if guard.held > 0 {
            guard.held -= 1;
            self.cond.notify_one();
        }
    }

    fn contention(&self) -> Contention {
        Contention::new("Failed to acquire semaphore within timeout")
    }
}

File: packages/lock/src/Semaphore.php
<?php

declare(strict_types=1);

namespace Utopia\Lock;

use InvalidArgumentException;
use Swoole\Coroutine;
use Swoole\Coroutine\Channel;
use Utopia\Lock\Exception\Contention;

final class Semaphore implements Lock
{
    private readonly Channel $channel;

    private int $syncHeld = 0;

    public function __construct(private readonly int $permits)
    {
        if ($permits < 1) {
            throw new InvalidArgumentException('Permits must be at least 1');
        }

        $this->channel = new Channel($permits);
    }

    #[\Override]
    public function acquire(float $timeout = 0.0): bool
    {
        if (! $this->inCoroutine()) {
            if ($this->syncHeld >= $this->permits) {
                return false;
            }
            $this->syncHeld++;

            return true;
        }

        return (bool) $this->channel->push(true, $timeout);
    }

    #[\Override]
    public function tryAcquire(): bool
    {
        if (! $this->inCoroutine()) {
            if ($this->syncHeld >= $this->permits) {
                return false;
            }
            $this->syncHeld++;

            return true;
        }

        if ($this->channel->isFull()) {
            return false;
        }

        return (bool) $this->channel->push(true, 0.001);
    }

    #[\Override]
    public function release(): void
    {
        if (! $this->inCoroutine()) {
            if ($this->syncHeld > 0) {
                $this->syncHeld--;
            }

            return;
        }

        if ($this->channel->isEmpty()) {
            return;
        }

        $this->channel->pop(0.001);
    }

    #[\Override]
    public function withLock(callable $callback, float $timeout = 0.0): mixed
    {
        if (! $this->acquire($timeout)) {
            throw new Contention('Failed to acquire semaphore within timeout');
        }

        try {
            return $callback();
        } finally {
            $this->release();
        }
    }

    private function inCoroutine(): bool
    {
        return \extension_loaded('swoole') && Coroutine::getCid() > 0;
    }
}
